internal/controller/promotions/promotions.go

```diff
@@ -486,3 +492,28 @@ func (r *reconciler) promote(
 
 	return newStatus, nil
 }
+
+// buildTargetFreightCollection constructs a FreightCollection that contains all
+// FreightReferences at the top of the Stage's FreightHistory stack (excepting
+// those that are no longer requested), plus a FreightReference for the provided
+// targetFreight.
+func (r *reconciler) buildTargetFreightCollection(
+	targetFreight kargoapi.FreightReference,
+	stage *kargoapi.Stage,
+) *kargoapi.FreightCollection {
+	freightCol := &kargoapi.FreightCollection{}
+
+	// We don't simply copy the current FreightCollection because we want to
+	// account for the possibility that some freight contained therein are no
+	// longer requested by the Stage.
```

Comment:
Should we be careful about the opposite thing potentially happening? I.e. there is now Freight requested for which we do not have an item in the stack (yet).

Reply:
That will happen, but I don't think there's anything special about that scenario as it is essentially no different than a _first-time_ promotion of one piece of Freight to a new Stage that requires n > 1. That's a scenario we had discussed and iirc, we agreed the likely result is reaching a state that will either not become healthy or will fail verification, but that that is also correct behavior for such a scenario. A subsequent promotion will set things right.

_But..._

> we agreed the likely result is reaching a state that will either not become healthy or will fail verification

I'm thinking now about scenarios where this may not be true -- perhaps, for instance, a scenario wherein the Promotion itself cannot succeed without all n Freight. In that case, the Stage's FreightHistory will remain forever empty, making it impossible for any promotion to ever provide the nth of n requested Freight as they will always be providing only the 1st of n.

I will noodle on this for a bit. Please lmk if you have ideas.

Reply:
Oh... if we built up the collection of Freight used by the Promotion (apart from the Freight referenced _by_ the Promotion) by getting it from `stage.Status.LastPromotion.Status.FreightCollection` _instead of_ `stage.Status.FreightHistory.Current().Freight` that may solve this. It would allow a second promotion, providing the 2nd of n > 1 Freight to build upon the failure of the first Promotion. For the average case where Promotions succeed, the overall effect of this change would still be identical to the current logic.

wdyt?

Reply:
I like this idea better, as it would allow you to "fill" Freight until your promotion succeeds.

Reply:
Ok. Will do this.

Reply:
@hiddeco, this is fixed in 73c5e8e00a3ac36e875d64b4cfd9ab7a36d3db75